service/history/nDCTransactionMgr.go

```diff
@@ -0,0 +1,232 @@
+)
+
+// NOTE: terminology
+//
+// 1. currentWorkflow means current running / closed workflow
+//  pointed by the current record in DB
+//
+// 2. targetWorkflow means the workflow to be replicated
+//  pointed by the replication task
+//
+// 3. newWorkflow means the workflow to be replicated, as part of continue as new
+//  pointed by the replication task
+//
+// 4. if target workflow and current workflow are the same
+//  then target workflow is set, current workflow is nil
+//
+// 5. suppress a workflow means turn a workflow into a zombie
+//  or terminate a workflow
+
+// Cases to be handled by this file:
+//
+// create path (there will be only current branch)
+// 1. create as current															-> nDCTransactionPolicyCreateAsCurrent
+// 2. create as zombie															-> nDCTransactionPolicyCreateAsZombie
+//
+// create path (there will be only current branch) + suppress current
+// 1. create as current & suppress current										-> nDCTransactionPolicySuppressCurrentAndCreateAsCurrent
+//
+// update to current branch path
+// 1. update as current															-> nDCTransactionPolicyUpdateAsCurrent
+// 2. update as current & new created as current								-> nDCTransactionPolicyUpdateAsCurrent
+// 3. update as zombie															-> nDCTransactionPolicyUpdateAsZombie
+// 4. update as zombie & new created as zombie									-> nDCTransactionPolicyUpdateAsZombie
+//
+// backfill to non current branch path
+// 1. backfill as is
+// 2. backfill as is & new created as zombie
+//
+// conflict resolve path
+// 1. conflict resolve as current												-> nDCTransactionPolicyConflictResolveAsCurrent
+// 2. conflict resolve as zombie												-> nDCTransactionPolicyConflictResolveAsZombie
+//
+// conflict resolve path + suppress current
+// 1. update from zombie to current & suppress current							-> nDCTransactionPolicySuppressCurrentAndUpdateAsCurrent
+// 2. update from zombie to current & new created as current & suppress current	-> nDCTransactionPolicySuppressCurrentAndUpdateAsCurrent
+
+type nDCTransactionPolicy int
+
+const (
+	nDCTransactionPolicyCreateAsCurrent nDCTransactionPolicy = iota
+	nDCTransactionPolicyCreateAsZombie
+	nDCTransactionPolicySuppressCurrentAndCreateAsCurrent
+
+	nDCTransactionPolicyUpdateAsCurrent
+	nDCTransactionPolicyUpdateAsZombie
+
+	nDCTransactionPolicyConflictResolveAsCurrent
+	nDCTransactionPolicyConflictResolveAsZombie
+
+	nDCTransactionPolicySuppressCurrentAndUpdateAsCurrent
+)
+
+type (
+	nDCTransactionMgr interface {
+		createWorkflow(
+			ctx ctx.Context,
+			now time.Time,
+			targetWorkflow nDCWorkflow,
+		) error
+		updateWorkflow(
+			ctx ctx.Context,
+			now time.Time,
+			isWorkflowRebuilt bool,
+			targetWorkflow nDCWorkflow,
+			newWorkflow nDCWorkflow,
+		) error
+		backfillWorkflow(
+			ctx ctx.Context,
+			now time.Time,
+			targetWorkflow nDCWorkflow,
+			targetWorkflowEvents *persistence.WorkflowEvents,
+		) error
+	}
+
+	nDCTransactionMgrImpl struct {
```

Comment:
So this is a nested layer like nDCTransactionMgrImpl contains nDCCreateTransactionMgrImpl and nDCCreateTransactionMgrImpl contains nDCTransactionMgrImpl?

Reply:
nDCTransactionMgrImpl contains the necessary common methods, for nDCCreateTransactionMgrImpl and nDUpdateTransactionMgrImpl.

nDCTransactionMgrImpl will do dispatching to nDCCreateTransactionMgrImpl and nDUpdateTransactionMgrImpl accordingly.

splitting nDCCreateTransactionMgrImpl and nDUpdateTransactionMgrImpl helps keeping the logics clean